core/src/main/java/tachyon/underfs/UnderFileSystemRegistry.java

```diff
@@ -0,0 +1,256 @@
+ * <p/>
+ * <h4>Automatic Discovery</h4>
+ * <p>
+ * To use the {@linkplain ServiceLoader} based mechanism you need to have a file named
+ * {@code tachyon.underfs.UnderFileSystemFactory} placed in the {@code META-INF\services} directory
+ * of your project. This file should contain the full name of your factory types (one per line),
+ * your factory types must have a public unparameterised constructor available (see
+ * {@link ServiceLoader} for more detail on this). You can enable {@code DEBUG} level logging to see
+ * factories as they are discovered if you wish to check that your implementation gets discovered.
+ * </p>
+ * <p>
+ * Note that if you are bundling Tachyon plus your code in a shaded JAR using Maven make sure to use
+ * the {@code ServicesResourceTransformer} as otherwise your services file will override the core
+ * provided services file and leave the standard factories and under file system implementations
+ * unavailable.
+ * </p>
+ * <h4>Manual Registration</h4>
+ * <p>
+ * To manually register a factory simply pass an instance of your factory to the
+ * {@link #register(UnderFileSystemFactory)} method. This can be useful when your factory cannot be
+ * instantiated without arguments or in cases where automatic discovery does not give your factory
+ * priority. Factories registered this way will be registered at the start of the factories list so
+ * will have the first opportunity to indicate whether they support a requested path.
+ * </p>
+ */
+public class UnderFileSystemRegistry {
+
+  private static final List<UnderFileSystemFactory> FACTORIES =
+      new CopyOnWriteArrayList<UnderFileSystemFactory>();
+  private static final Logger LOG = LoggerFactory.getLogger(Constants.LOGGER_TYPE);
+  private static boolean sInit = false;
+
+  static {
+    // Call the actual initializer which is a synchronized method for thread safety purposes
+    init();
+  }
+
+  /**
+   * Returns a read-only view of the available factories
+   * 
+   * @return Read-only view of the available factories
+   */
+  public static List<UnderFileSystemFactory> available() {
+    return Collections.unmodifiableList(FACTORIES);
+  }
+
+  /**
+   * Creates a client that can talk to the under file system
+   * 
+   * @param path Path
+   * @param conf Optional configuration object for the UFS, may be null
+   * @param tachyonConf Tachyon Configuration
+   * @return Client for the under file system
+   * @throws IllegalArgumentException Thrown if there is no under file system for the given path or
+   *         if no under file system could successfully be created
+   */
+  public static UnderFileSystem create(String path, TachyonConf tachyonConf, Object conf) {
+    // Try to obtain the appropriate factory
+    List<UnderFileSystemFactory> factories = findAll(path, tachyonConf);
+    if (factories.size() == 0) {
+      throw new IllegalArgumentException("No known Under File System supports the given path "
```

Comment:
This case is when no registered ufs supports the path, whereas below one or more ufs have stated they support the path but failed to support it due to some error.